### DISCUSSION TITLE
Narrowing the union of `Any` and generic with structural pattern matching gives unknown types

### DISCUSSION THREAD

**Discussion by sylee957** [microsoft/pyright#4864] [Q&A]
I have an example with a generic `Identity` and a union type `Any | Identity[T]`

```python3
from dataclasses import dataclass
from typing import Any, Generic, TypeVar, reveal_type

T = TypeVar("T")


@dataclass
class Identity(Generic[T]):
    value: T
```

If instance check is used, `value` is narrowed down to `T`
```python3
def test(t: Any | Identity[T]):
    if isinstance(t, Identity):
        reveal_type(t.value)
    else:
        reveal_type(t)
```

However, if structural pattern matching is used, `value` is narrowed down to `Unknown | T`
```python3
def test_1(t: Any | Identity[T]):
    match t:
        case Identity(value):
            reveal_type(value)
        case _:
            reveal_type(t)
```
which seems like some inconsistency.

**Comment by erictraut** [comment]
Yes, the `isinstance` narrowing logic is completely separate from the narrowing logic used in structural pattern matching. There will naturally be some difference between the two. While the two mechanisms (`isinstance` checks and class patterns) are conceptually similar, they are quite different in how they work.

I'll take a closer look and see if I can reduce the inconsistency between the two cases without regressing anything.

**Reply by erictraut** [reply]
Taking a closer look at this, I think the narrowing logic for class patterns is more correct than the existing `isinstance` narrowing logic. Both should result in `Unknown` types in your sample above.

I made the change to the `isinstance` logic, but this created a bunch of noise in existing strictly-typed code bases that use pyright, so I think such a change would not be well received by the pyright community, even though it's arguably "more correct" than the current behavior.

Rather than propagate this less-correct behavior to the class pattern narrowing logic, I'm going to leave everything as it is currently.
